### PR TITLE
Add encoding & intrinsic for VR -> GPR ALU ops

### PR DESCRIPTION
**Pull request by pranav-prakash**
Sorry for the delay, I'd been busy with finals. This patch adds the encoding and templates for alu operations such as vpopcnt that take a single vector register and write to a gpr.

### CODE REVIEW

File: lib/Target/RISCV/RISCVInstrFormatsV.td [hunk 1]
```diff
@@ -19,7 +19,7 @@ class RVVMaskCond<bits<2> m> {
 
 def RVV_Unmasked : RVVMaskCond<0b01>;
 
-class RVInstVR<bits<7> funct7, bits<1> funct1, RVVMaskCond m, RISCVOpcode opcode,
+class RVInstVrVr<bits<7> funct7, bits<1> funct1, RVVMaskCond m, RISCVOpcode opcode,
```
Comment (rkruppe) [inline lib/Target/RISCV/RISCVInstrFormatsV.td]: I don't know how I feel about this name, but it's not important since eventually we'll just use whatever name the spec uses for this format. But please do add a comment similar to the one on `RVInstVrGpr` for completeness.

File: lib/Target/RISCV/RISCVInstrFormatsV.td [hunk 2]
```diff
@@ -35,6 +35,22 @@ class RVInstVR<bits<7> funct7, bits<1> funct1, RVVMaskCond m, RISCVOpcode opcode
   let Opcode = opcode.Value;
 }
 
+//Vector -> GPR 
+class RVInstVrGpr<bits<3> funct3, bits<5> funct5, RVVMaskCond m, RISCVOpcode opcode,
```
Comment (rkruppe) [inline lib/Target/RISCV/RISCVInstrFormatsV.td]: All this appears to be from the recent encoding proposal for 0.6-draft. I'd prefer to use the old encoding for now and switch everything over to the new encoding at once, both to avoid confusion (I was staring at the old encoding table for a minute or two before I realize this is the new encoding) and because there is a risk of encoding conflicts if half the instructions use a different encoding scheme than the other half.

File: lib/Target/RISCV/RISCVOptimizeVSETVLUses.cpp [hunk 3]
```diff
@@ -65,6 +65,9 @@ bool RISCVOptimizeVSETVLUses::runOnMachineFunction(MachineFunction &Fn) {
           const auto& CopyDest = Instr.getOperand(0);
           auto& CopySource = Instr.getOperand(1);
 
+          if (!TargetRegisterInfo::isVirtualRegister(CopySource.getReg()))
```
Comment (rkruppe) [inline lib/Target/RISCV/RISCVOptimizeVSETVLUses.cpp]: Why is this necessary (and why wasn't it necessary before)? Could you add a comment explaining that?

File: lib/Target/RISCV/RISCVInstrFormatsV.td (hunk 1 above)
Comment (pranav-prakash) [reply lib/Target/RISCV/RISCVInstrFormatsV.td]: Done

File: lib/Target/RISCV/RISCVInstrFormatsV.td (hunk 2 above)
Comment (pranav-prakash) [reply lib/Target/RISCV/RISCVInstrFormatsV.td]: Yeah I was using the newer instruction encoding. I've switched it back to the older one from the 0.5 spec.

File: lib/Target/RISCV/RISCVOptimizeVSETVLUses.cpp (hunk 3 above)
Comment (pranav-prakash) [reply lib/Target/RISCV/RISCVOptimizeVSETVLUses.cpp]: If the COPY source is a live-in (which I think are things passed as parameters?) then `getVRegDef` returns a nullptr which causes a segfault when it's dereferenced later on. It should have been included before as well, but I only found this potential issue when writing the test case for `vmpopc`.

Comment (rkruppe) [reply lib/Target/RISCV/RISCVOptimizeVSETVLUses.cpp]: Ok, this makes sense, but it would be more robust and IMO clearer if the code just checked if `getVRegDef` returns null:

```c++
const MachineInstr *MI = MRI.getVRegDef(CopySource.getReg());
if (!MI)
    // No definition (e.g., a live-in physical register), can't optimize
    continue;
```

Comment (pranav-prakash) [reply lib/Target/RISCV/RISCVOptimizeVSETVLUses.cpp]: Done!

**Comment by rkruppe** [comment]
Thanks!
